src/main/java/eu/arrowhead/client/library/ArrowheadService.java

```diff
@@ -453,6 +468,79 @@ public OrchestrationResponseDTO queryOrchestrationStore(final long systemId) {
+	 * @param token (nullable) String value which represents the token for being authorized at the provider side if necessary. Token could be received in orchestration response per interface type.
+	 * @param queryParams (nullable) String... variable arguments which represent the additional key-value http(s) query parameters if any necessary. E.g.: "k1", "v1", "k2", "v2".
+	 * @return A WebSocketConnection manager that is used for all full-duplex communication
+	 *
+	 * @throws InvalidParameterException when service URL can't be assembled.
+	 * @throws CertificateException when a certificate or SSL error occurs
+	 * @throws ArrowheadException when the communication is managed via Gateway Core System and internal server error happened.
+	 */
+	public WebSocketConnectionManager connnectServiceWS(final WebSocketHandler handler, final String address, final int port, final String serviceUri,
+														final String token, final String... queryParams) {
+
+		if (Utilities.isEmpty(address)) {
+			throw new InvalidParameterException("address cannot be null or blank.");
+		}
+		if (Utilities.isEmpty(serviceUri)) {
+			throw new InvalidParameterException("serviceUri cannot be null or blank.");
+		}
+		if (port < 0 || port > 65535){
+			throw new InvalidParameterException("port cannot be negative or larger that 65535");
+		}
+
+		/* Handle query parameters */
+		String[] validatedQueryParams;
+		if (queryParams == null) {
+			validatedQueryParams = new String[0];
+		} else {
+			validatedQueryParams = queryParams;
+		}
+
+		String wsSec = "ws";
+		if( sslProperties.isSslEnabled() ) {
+			wsSec ="wss";
+		}
+
+		/* prepare the URI */
+		UriComponents uri;
+		if(!Utilities.isEmpty(token)) {
+			final List<String> query = new ArrayList<>();
+			query.addAll(Arrays.asList(validatedQueryParams));
+			query.add(CommonConstants.REQUEST_PARAM_TOKEN);
+			query.add(token);
+			uri = Utilities.createURI(wsSec, address, port, serviceUri, query.toArray(new String[query.size()]));
+		} else {
+			uri = Utilities.createURI(wsSec, address, port, serviceUri, validatedQueryParams);
+		}
+
+		/* try to establish WS(S) connection */
+		final StandardWebSocketClient wsClient = new StandardWebSocketClient();
+		if(sslProperties.isSslEnabled()) {
+			try {
+				final KeyStore trustStore = getTrustStore(); 
+				final KeyStore keyStore = getKeyStore();
+				final TrustStrategy acceptingTrustStrategy = (X509Certificate[] chain, String authType) -> true;
+				final SSLContext sslContext = SSLContexts.custom().loadTrustMaterial(trustStore, acceptingTrustStrategy).loadKeyMaterial(keyStore, sslProperties.getKeyStorePassword().toCharArray()).build();
+
+				wsClient.getUserProperties().clear();
+				wsClient.getUserProperties().put(ClientCommonConstants.TOMCAT_WS_SSL_CONTEXT, sslContext);
+			} catch(Exception e) {
+				throw new ArrowheadException("WSS connection failed: " + e.getMessage());
+			}
+		}
+		
+		return new WebSocketConnectionManager(wsClient, handler, uri.toString());
```

Comment:
I think `WebSocketConnectionManager.start()` should be called before the return and the method could be void, because the given `WebSocketHandler ` implements the actual client related business logic. What is your opinion?

Reply:
Fixed